executor/src/wasm_executor.rs

```diff
@@ -77,16 +78,30 @@ impl WritePrimitive<u32> for MemoryInstance {
 }
 
 impl_function_executor!(this: FunctionExecutor<'e, E>,
-	ext_print(utf8_data: *const u8, utf8_len: u32) => {
+	ext_print_utf8(utf8_data: *const u8, utf8_len: u32) => {
 		if let Ok(utf8) = this.memory.get(utf8_data, utf8_len as usize) {
 			if let Ok(message) = String::from_utf8(utf8) {
 				println!("Runtime: {}", message);
 			}
 		}
 	},
+	ext_print_hex(data: *const u8, len: u32) => {
+		if let Ok(hex) = this.memory.get(data, len as usize) {
+			println!("Runtime: {}", HexDisplay::from(&hex));
+		}
+	},
 	ext_print_num(number: u64) => {
 		println!("Runtime: {}", number);
 	},
+	ext_memcmp(s1: *const u8, s2: *const u8, n: usize) -> i32 => {
+		if let (Ok(sl1), Ok(sl2))
+			= (this.memory.get(s1, n as usize), this.memory.get(s2, n as usize)) {
+			use memcmp::Memcmp;
+			(&sl1).memcmp(&sl2) as i32
+		} else {
+			0
+		}
+	},
```

Comment:
`memcmp` should return `0` if `sl1` == `sl2`, i.e. all bytes are equal,
should return a negative value if first different byte in `sl1` is less than byte at the same position in `sl2` and should return a positive value if this different byte from `sl1` is greater than corresponding byte from `sl2`.

The problem is `memcmp::Memcmp` for some reason returns `bool` value, and seems like it returns `true` if slices are equal. So this implementation returns `1` in case slices are equal and `0` in other cases.

Also, in case if `s1` or `s2` can't be read, you return zero. This is not strictly a problem, because passing `s1` or `s2` here is undefined behavior and for such case returning `0` is perfeclty legal.
However, I strongly recommend to trap here, because:

1. returning some value, especially `0` (meaning that `s1` == `s2`), might lead to serious security problems,
1. such behavior will be inline with behavior of `memcmp` implemented within wasm itself.

Here is [test cases](https://github.com/alexcrichton/rlibc/blob/master/src/lib.rs#L96) that might help you to test this `memcmp` extern.